# Identifying Features Influencing Value of Homes in Ames, Iowa

## Introduction

In this project, we wish to identify features of homes in Ames, Iowa which would strongly influence its value. These information would benefit homeowners who are looking to evaluate their current living environment while increasing their ability to negotiate for a higher price in a future sales. Homeseekers would also do well to note the particular presence or absence of certain features to make estimation of the price of the home they desire.

This analysis will be carried out by using a rich dataset of housing prices in Ames, Iowa to create a model for predicting sale price of a given home in the city. At the same time, the model chosen would quantify how the absence, presence and size of a particular feature affect the final selling price, enabling the homeowner to make cost benefit analysis on the type of enhancement in their homes.

## Data Import and Cleaning

### Import the libraries

In [1]:
import pandas as pd
import numpy as np


In [2]:
pd.set_option("display.max_rows", 100)
pd.set_option("display.max_columns", 100)

### Import the train and test dataset

In [3]:
df_train = pd.read_csv('../data/train.csv')
df_test = pd.read_csv('../data/new_test.csv')

In [4]:
df_test.shape

(879, 80)

In [5]:
df_train.shape

(2051, 81)

In [6]:
df_train.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley  \
0  109  533352170           60        RL           NaN     13517   Pave   NaN   
1  544  531379050           60        RL          43.0     11492   Pave   NaN   
2  153  535304180           20        RL          68.0      7922   Pave   NaN   
3  318  916386060           60        RL          73.0      9802   Pave   NaN   
4  255  906425045           50        RL          82.0     14235   Pave   NaN   

  Lot Shape Land Contour Utilities Lot Config Land Slope Neighborhood  \
0       IR1          Lvl    AllPub    CulDSac        Gtl       Sawyer   
1       IR1          Lvl    AllPub    CulDSac        Gtl      SawyerW   
2       Reg          Lvl    AllPub     Inside        Gtl        NAmes   
3       Reg          Lvl    AllPub     Inside        Gtl       Timber   
4       IR1          Lvl    AllPub     Inside        Gtl      SawyerW   

  Condition 1 Condition 2 Bldg Type House Style  Overall Qual  Overall Cond  \
0        RRAe        Norm      1Fam      2Story             6             8   
1        Norm        Norm      1Fam      2Story             7             5   
2        Norm        Norm      1Fam      1Story             5             7   
3        Norm        Norm      1Fam      2Story             5             5   
4        Norm        Norm      1Fam      1.5Fin             6             8   

   Year Built  Year Remod/Add Roof Style Roof Matl Exterior 1st Exterior 2nd  \
0        1976            2005      Gable   CompShg      HdBoard      Plywood   
1        1996            1997      Gable   CompShg      VinylSd      VinylSd   
2        1953            2007      Gable   CompShg      VinylSd      VinylSd   
3        2006            2007      Gable   CompShg      VinylSd      VinylSd   
4        1900            1993      Gable   CompShg      Wd Sdng      Plywood   

  Mas Vnr Type  Mas Vnr Area Exter Qual Exter Cond Foundation Bsmt Qual  \
0      BrkFace         289.0         Gd         TA     CBlock        TA   
1      BrkFace         132.0         Gd         TA      PConc        Gd   
2         None           0.0         TA         Gd     CBlock        TA   
3         None           0.0         TA         TA      PConc        Gd   
4         None           0.0         TA         TA      PConc        Fa   

  Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1 BsmtFin Type 2  \
0        TA            No            GLQ         533.0            Unf   
1        TA            No            GLQ         637.0            Unf   
2        TA            No            GLQ         731.0            Unf   
3        TA            No            Unf           0.0            Unf   
4        Gd            No            Unf           0.0            Unf   

   BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF Heating Heating QC Central Air  \
0           0.0        192.0          725.0    GasA         Ex           Y   
1           0.0        276.0          913.0    GasA         Ex           Y   
2           0.0        326.0         1057.0    GasA         TA           Y   
3           0.0        384.0          384.0    GasA         Gd           Y   
4           0.0        676.0          676.0    GasA         TA           Y   

  Electrical  1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  \
0      SBrkr         725         754                0         1479   
1      SBrkr         913        1209                0         2122   
2      SBrkr        1057           0                0         1057   
3      SBrkr         744         700                0         1444   
4      SBrkr         831         614                0         1445   

   Bsmt Full Bath  Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  \
0             0.0             0.0          2          1              3   
1             1.0             0.0          2          1              4   
2             1.0             0.0          1          0              3   
3             0.0             0.0          2  

In [7]:
df_test.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0  2658  902301120          190        RM          69.0      9142   Pave   
1  2718  905108090           90        RL           NaN      9662   Pave   
2  2414  528218130           60        RL          58.0     17104   Pave   
3  1989  902207150           30        RM          60.0      8520   Pave   
4   625  535105100           20        RL           NaN      9500   Pave   

  Alley Lot Shape Land Contour Utilities Lot Config Land Slope Neighborhood  \
0  Grvl       Reg          Lvl    AllPub     Inside        Gtl      OldTown   
1   NaN       IR1          Lvl    AllPub     Inside        Gtl       Sawyer   
2   NaN       IR1          Lvl    AllPub     Inside        Gtl      Gilbert   
3   NaN       Reg          Lvl    AllPub     Inside        Gtl      OldTown   
4   NaN       IR1          Lvl    AllPub     Inside        Gtl        NAmes   

  Condition 1 Condition 2 Bldg Type House Style  Overall Qual  Overall Cond  \
0        Norm        Norm    2fmCon      2Story             6             8   
1        Norm        Norm    Duplex      1Story             5             4   
2        Norm        Norm      1Fam      2Story             7             5   
3        Norm        Norm      1Fam      1Story             5             6   
4        Norm        Norm      1Fam      1Story             6             5   

   Year Built  Year Remod/Add Roof Style Roof Matl Exterior 1st Exterior 2nd  \
0        1910            1950      Gable   CompShg      AsbShng      AsbShng   
1        1977            1977      Gable   CompShg      Plywood      Plywood   
2        2006            2006      Gable   CompShg      VinylSd      VinylSd   
3        1923            2006      Gable   CompShg      Wd Sdng      Wd Sdng   
4        1963            1963      Gable   CompShg      Plywood      Plywood   

  Mas Vnr Type  Mas Vnr Area Exter Qual Exter Cond Foundation Bsmt Qual  \
0         None           0.0         TA         Fa      Stone        Fa   
1         None           0.0         TA         TA     CBlock        Gd   
2         None           0.0         Gd         TA      PConc        Gd   
3         None           0.0         Gd         TA     CBlock        TA   
4      BrkFace         247.0         TA         TA     CBlock        Gd   

  Bsmt Cond Bsmt Exposure BsmtFin Type 1  BsmtFin SF 1 BsmtFin Type 2  \
0        TA            No            Unf             0            Unf   
1        TA            No            Unf             0            Unf   
2        Gd            Av            GLQ           554            Unf   
3        TA            No            Unf             0            Unf   
4        TA            No            BLQ           609            Unf   

   BsmtFin SF 2  Bsmt Unf SF  Total Bsmt SF Heating Heating QC Central Air  \
0             0         1020           1020    GasA         Gd           N   
1             0         1967           1967    GasA         TA           Y   
2             0          100            654    GasA         Ex           Y   
3             0          968            968    GasA         TA           Y   
4             0          785           1394    GasA         Gd           Y   

  Electrical  1st Flr SF  2nd Flr SF  Low Qual Fin SF  Gr Liv Area  \
0      FuseP         908        1020                0         1928   
1      SBrkr        1967           0                0         1967   
2      SBrkr         664         832                0         1496   
3      SBrkr         968           0                0          968   
4      SBrkr        1394           0                0         1394   

   Bsmt Full Bath  Bsmt Half Bath  Full Bath  Half Bath  Bedroom AbvGr  \
0               0               0          2          0              4   
1               0               0          2          0              6   
2               1               0          2          1              3   
3               0               0        

### Data Dictionary

Please refer to this [link](http://jse.amstat.org/v19n3/decock/DataDocumentation.txt) for the variable description and categories.

### Rename the columns

In [8]:
df_train.columns=df_train.columns.map(lambda col_name: col_name.lower().replace(' ', '_').replace('/', '_'))
df_test.columns=df_test.columns.map(lambda col_name: col_name.lower().replace(' ', '_').replace('/', '_'))
df_train.columns

Index(['id', 'pid', 'ms_subclass', 'ms_zoning', 'lot_frontage', 'lot_area',
       'street', 'alley', 'lot_shape', 'land_contour', 'utilities',
       'lot_config', 'land_slope', 'neighborhood', 'condition_1',
       'condition_2', 'bldg_type', 'house_style', 'overall_qual',
       'overall_cond', 'year_built', 'year_remod_add', 'roof_style',
       'roof_matl', 'exterior_1st', 'exterior_2nd', 'mas_vnr_type',
       'mas_vnr_area', 'exter_qual', 'exter_cond', 'foundation', 'bsmt_qual',
       'bsmt_cond', 'bsmt_exposure', 'bsmtfin_type_1', 'bsmtfin_sf_1',
       'bsmtfin_type_2', 'bsmtfin_sf_2', 'bsmt_unf_sf', 'total_bsmt_sf',
       'heating', 'heating_qc', 'central_air', 'electrical', '1st_flr_sf',
       '2nd_flr_sf', 'low_qual_fin_sf', 'gr_liv_area', 'bsmt_full_bath',
       'bsmt_half_bath', 'full_bath', 'half_bath', 'bedroom_abvgr',
       'kitchen_abvgr', 'kitchen_qual', 'totrms_abvgrd', 'functional',
       'fireplaces', 'fireplace_qu', 'garage_type', 'garage_yr_blt',
       'g

In [9]:
df_test.columns

Index(['id', 'pid', 'ms_subclass', 'ms_zoning', 'lot_frontage', 'lot_area',
       'street', 'alley', 'lot_shape', 'land_contour', 'utilities',
       'lot_config', 'land_slope', 'neighborhood', 'condition_1',
       'condition_2', 'bldg_type', 'house_style', 'overall_qual',
       'overall_cond', 'year_built', 'year_remod_add', 'roof_style',
       'roof_matl', 'exterior_1st', 'exterior_2nd', 'mas_vnr_type',
       'mas_vnr_area', 'exter_qual', 'exter_cond', 'foundation', 'bsmt_qual',
       'bsmt_cond', 'bsmt_exposure', 'bsmtfin_type_1', 'bsmtfin_sf_1',
       'bsmtfin_type_2', 'bsmtfin_sf_2', 'bsmt_unf_sf', 'total_bsmt_sf',
       'heating', 'heating_qc', 'central_air', 'electrical', '1st_flr_sf',
       '2nd_flr_sf', 'low_qual_fin_sf', 'gr_liv_area', 'bsmt_full_bath',
       'bsmt_half_bath', 'full_bath', 'half_bath', 'bedroom_abvgr',
       'kitchen_abvgr', 'kitchen_qual', 'totrms_abvgrd', 'functional',
       'fireplaces', 'fireplace_qu', 'garage_type', 'garage_yr_blt',
       'g

### Review dtype and null values

In [10]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2051 entries, 0 to 2050
Data columns (total 81 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               2051 non-null   int64  
 1   pid              2051 non-null   int64  
 2   ms_subclass      2051 non-null   int64  
 3   ms_zoning        2051 non-null   object 
 4   lot_frontage     1721 non-null   float64
 5   lot_area         2051 non-null   int64  
 6   street           2051 non-null   object 
 7   alley            140 non-null    object 
 8   lot_shape        2051 non-null   object 
 9   land_contour     2051 non-null   object 
 10  utilities        2051 non-null   object 
 11  lot_config       2051 non-null   object 
 12  land_slope       2051 non-null   object 
 13  neighborhood     2051 non-null   object 
 14  condition_1      2051 non-null   object 
 15  condition_2      2051 non-null   object 
 16  bldg_type        2051 non-null   object 
 17  house_style   

In [11]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 879 entries, 0 to 878
Data columns (total 80 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               879 non-null    int64  
 1   pid              879 non-null    int64  
 2   ms_subclass      879 non-null    int64  
 3   ms_zoning        879 non-null    object 
 4   lot_frontage     719 non-null    float64
 5   lot_area         879 non-null    int64  
 6   street           879 non-null    object 
 7   alley            58 non-null     object 
 8   lot_shape        879 non-null    object 
 9   land_contour     879 non-null    object 
 10  utilities        879 non-null    object 
 11  lot_config       879 non-null    object 
 12  land_slope       879 non-null    object 
 13  neighborhood     879 non-null    object 
 14  condition_1      879 non-null    object 
 15  condition_2      879 non-null    object 
 16  bldg_type        879 non-null    object 
 17  house_style     

### Custom Functions and Variable Type Lists

#### Create Function for showing variables with null values and sum of null values in a DataFrame and for a variable type 

In [12]:
def var_null_count(df, var_list=None):
    '''This function takes in a dataframe df and an optional variable type list var_type_list 
    and returns a series null_count containing only variables in the dataframe that has 
    null values and the number of null values'''
    
    null = df.isna().sum()
    if var_list is None:
        null_count = null[null!=0]       
    else:
        null_count=null[(null!=0)&(null.index.isin(var_list))]
                                                            
    return null_count

#### Create Function for showing dictionary of unique values of columns in a DataFrame and for a variable type

In [13]:
def unique_val_dict(df, var_type_list):
    '''This function takes in a dataframe df and an variable type list and returns a dict unique_val containing 
    unique values of columns in the dataframe for a variable type'''
    null = df.isna().sum()
    null_count=null[(null!=0)&(null.index.isin(var_type_list))]
    unique_val = {col: df[col].unique() for col in null_count.index}
    return unique_val

#### Create lists of continous, nominal, ordinal and discrete variables based on the data description

In [14]:
# List of continuous variables
cont_var = ['lot_frontage', 'bsmtfin_sf_1', 'bsmtfin_sf_2', 'bsmt_unf_sf', 'total_bsmt_sf', 'garage_area', 
            'mas_vnr_area', 'lot_area', '1st_flr_sf', '2nd_flr_sf', 'low_qual_fin_sf', 'gr_liv_area', 
            'wood_deck_sf', 'open_porch_sf', 'enclosed_porch', '3ssn_porch', 'screen_porch', 'pool_area', 
            'misc_val','saleprice']

# List of nominal variables
nom_var = ['id', 'pid','alley', 'mas_vnr_type', 'garage_type' , 'misc_feature', 'ms_subclass', 'ms_zoning', 
           'street', 'land_contour', 'lot_config', 'neighborhood', 'condition_1', 'condition_2', 'bldg_type', 
           'house_style', 'roof_style', 'roof_matl', 'exterior_1st', 'exterior_2nd', 'foundation', 'heating', 
           'central_air', 'sale_type']

# List of ordinal variables
ord_var = ['bsmt_qual', 'bsmt_cond', 'bsmt_exposure', 'bsmtfin_type_1', 'bsmtfin_type_2', 'fireplace_qu', 
           'garage_finish', 'garage_qual', 'garage_cond','pool_qc', 'fence', 'lot_shape', 'utilities', 
           'land_slope', 'overall_qual', 'overall_cond', 'heating_qc', 'electrical', 'kitchen_qual', 
           'functional', 'paved_drive','exter_qual', 'exter_cond']

# List of discrete variables
disc_var = ['bsmt_full_bath', 'bsmt_half_bath', 'garage_yr_blt', 'garage_cars', 'year_built', 'year_remod/add',
            'full_bath', 'half_bath', 'bedroom_abvgr', 'kitchen_abvgr', 'totrms_abvgrd', 'fireplaces', 'mo_sold', 
            'yr_sold'] 

### Review null values in nominal variables

In [15]:
# Display all unique categories in nominal variables with null values 
unique_val_dict(df_train, nom_var)

{'alley': array([nan, 'Pave', 'Grvl'], dtype=object),
 'mas_vnr_type': array(['BrkFace', 'None', nan, 'Stone', 'BrkCmn'], dtype=object),
 'garage_type': array(['Attchd', 'Detchd', 'BuiltIn', 'Basment', nan, '2Types', 'CarPort'],
       dtype=object),
 'misc_feature': array([nan, 'Shed', 'TenC', 'Gar2', 'Othr', 'Elev'], dtype=object)}

In [16]:
unique_val_dict(df_test, nom_var)

{'alley': array(['Grvl', nan, 'Pave'], dtype=object),
 'mas_vnr_type': array(['None', 'BrkFace', 'Stone', 'BrkCmn', 'CBlock', nan], dtype=object),
 'garage_type': array(['Detchd', 'Attchd', 'BuiltIn', nan, '2Types', 'CarPort', 'Basment'],
       dtype=object),
 'misc_feature': array([nan, 'Shed', 'Othr', 'Gar2'], dtype=object)}

**Key Observations:**

From the data dictionary, we understand the following:
1. The NaNs in `alley` correspond to the property having no alley access.
2. The NaNs in `mas_vnr_type` correspond to no mansonry veneer on the property. There are also some properties which are labeled as having 'None' for `mas_vnr_type`. 
3. The NaNs in `garage_type` correspond to no garage on the property.
4. The NaNs in `misc_feature` correspond to absence of miscellaneous features on the property.

#### Fix the null values in nominal variables

`alley` is a standalone variable which the null values can be replaced easily without considering other variables

In [17]:
# Replace null values on alley columns with 'NA'
df_train['alley'].fillna('NA', inplace=True)
df_test['alley'].fillna('NA', inplace=True)

Check how many properties with `mas_vnr_type` == None has `mas_vnr_area` == 0

In [18]:
mas_type_count=df_train.loc[(df_train['mas_vnr_type']=='None'),:].shape[0]
mas_area_count=df_train.loc[(df_train['mas_vnr_area']==0),:].shape[0]
mas_type_area_count = df_train.loc[(df_train['mas_vnr_type']=='None')&(df_train['mas_vnr_area']==0),:].shape[0]

print(f'There are {mas_type_count} properties with mas_vnr_type is None')
print(f'There are {mas_area_count} properties with mas_vnr_area is 0')
print(f'There are {mas_type_area_count} properties with mas_vnr_type is None and mas_vnr_area is 0')

There are 1218 properties with mas_vnr_type is None
There are 1216 properties with mas_vnr_area is 0
There are 1213 properties with mas_vnr_type is None and mas_vnr_area is 0


In [19]:
mas_type_count=df_test.loc[(df_test['mas_vnr_type']=='None'),:].shape[0]
mas_area_count=df_test.loc[(df_test['mas_vnr_area']==0),:].shape[0]
mas_type_area_count = df_test.loc[(df_test['mas_vnr_type']=='None')&(df_test['mas_vnr_area']==0),:].shape[0]

print(f'There are {mas_type_count} properties with mas_vnr_type is None')
print(f'There are {mas_area_count} properties with mas_vnr_area is 0')
print(f'There are {mas_type_area_count} properties with mas_vnr_type is None and mas_vnr_area is 0')

There are 534 properties with mas_vnr_type is None
There are 532 properties with mas_vnr_area is 0
There are 532 properties with mas_vnr_type is None and mas_vnr_area is 0


Show all properties with `mas_vnr_type` == None but `mas_vnr_area` != 0 and vice versa

In [20]:
df_train.loc[(df_train['mas_vnr_type']=='None')&(df_train['mas_vnr_area']!=0), ['mas_vnr_type', 'mas_vnr_area']]

mas_vnr_type  mas_vnr_area
765          None           1.0
810          None         288.0
1148         None           1.0
1684         None           1.0
1832         None         344.0

In [21]:
df_train.loc[(df_train['mas_vnr_type']!='None')&(df_train['mas_vnr_area']==0), ['mas_vnr_type', 'mas_vnr_area']]

mas_vnr_type  mas_vnr_area
669       BrkFace           0.0
1136      BrkFace           0.0
1187        Stone           0.0

In [22]:
df_test.loc[(df_test['mas_vnr_type']=='None')&(df_test['mas_vnr_area']!=0), ['mas_vnr_type', 'mas_vnr_area']]

mas_vnr_type  mas_vnr_area
383         None         285.0
863         None         312.0

**Key Observations:**

1. There may be properties where masonry veneer area is available but the properties chooses not to have any masonry veneer
2. The properties that has a masonry veneer type but with zero area may be a case of missing completely at random and will be removed

In [23]:
df_train.drop(axis=0, index=[669, 1136, 1187], inplace=True)

   Show all properties with both `mas_vnr_type` and `mas_vnr_area` are null values

In [24]:
df_train.loc[(df_train['mas_vnr_type'].isna()),['mas_vnr_type', 'mas_vnr_area']]

mas_vnr_type  mas_vnr_area
22            NaN           NaN
41            NaN           NaN
86            NaN           NaN
212           NaN           NaN
276           NaN           NaN
338           NaN           NaN
431           NaN           NaN
451           NaN           NaN
591           NaN           NaN
844           NaN           NaN
913           NaN           NaN
939           NaN           NaN
1025          NaN           NaN
1244          NaN           NaN
1306          NaN           NaN
1430          NaN           NaN
1434          NaN           NaN
1606          NaN           NaN
1699          NaN           NaN
1815          NaN           NaN
1820          NaN           NaN
1941          NaN           NaN

In [25]:
df_test.loc[(df_test['mas_vnr_type'].isna()),['mas_vnr_type', 'mas_vnr_area']]

mas_vnr_type  mas_vnr_area
866          NaN           NaN

Replace null values on `mas_vnr_type` columns with 'None' and replace null values in `mas_vnr_area` columns with zero

In [26]:
df_train['mas_vnr_type'].fillna('None', inplace=True)
df_train['mas_vnr_area'].fillna(0, inplace=True)

In [27]:
df_test['mas_vnr_type'].fillna('None', inplace=True)
df_test['mas_vnr_area'].fillna(0, inplace=True)

In [28]:
var_null_count(df_train, nom_var)

garage_type      113
misc_feature    1983
dtype: int64

In [29]:
var_null_count(df_test, nom_var)

garage_type      44
misc_feature    838
dtype: int64

Count number of properties where `garage_type`, `garage_yr_blt`, `garage_finish`, `garage_qual` and `garage_cond` are all null

In [30]:
garage_count = df_train.loc[((((df_train['garage_type'].isna())&
                            (df_train['garage_yr_blt'].isna()))&
                            (df_train['garage_finish'].isna()))&
                            (df_train['garage_qual'].isna()))&
                            (df_train['garage_cond'].isna())].shape[0] 


print(f'There are {garage_count} properties where "garage_type", "garage_yr_blt", "garage_finish", "garage_qual" and "garage_cond" are all null')

There are 113 properties where "garage_type", "garage_yr_blt", "garage_finish", "garage_qual" and "garage_cond" are all null


In [31]:
garage_count = df_test.loc[((((df_test['garage_type'].isna())&
                            (df_test['garage_yr_blt'].isna()))&
                            (df_test['garage_finish'].isna()))&
                            (df_test['garage_qual'].isna()))&
                            (df_test['garage_cond'].isna())].shape[0] 
                            

print(f'There are {garage_count} properties where "garage_type", "garage_yr_blt", "garage_finish", "garage_qual" and "garage_cond" are all null')

There are 44 properties where "garage_type", "garage_yr_blt", "garage_finish", "garage_qual" and "garage_cond" are all null


Replace all null values in `garage_type`, `garage_yr_blt`, `garage_finish`, `garage_qual` and `garage_cond` with 'NA', since these are most likely properties with no garage

In [32]:
garage_list = ['garage_type',
              'garage_yr_blt', 
              'garage_finish', 
              'garage_qual',
              'garage_cond']

var_null_count(df_train, garage_list)

garage_type      113
garage_yr_blt    114
garage_finish    114
garage_qual      114
garage_cond      114
dtype: int64

In [33]:
garage_list = ['garage_type',
              'garage_yr_blt', 
              'garage_finish', 
              'garage_qual',
              'garage_cond']

var_null_count(df_test, garage_list)

garage_type      44
garage_yr_blt    45
garage_finish    45
garage_qual      45
garage_cond      45
dtype: int64

In [34]:
for col_name in garage_list:
    df_train[col_name].fillna('NA', inplace=True)

In [35]:
for col_name in garage_list:
    df_test[col_name].fillna('NA', inplace=True)

Replace null values on misc features columns with 'NA' as it is a standalone variable which can be replaced easily without considering other variables

In [36]:
# Replace null values on misc features columns with 'NA'
df_train['misc_feature'].fillna('NA', inplace=True)

In [37]:
# Replace null values on misc features columns with 'NA'
df_test['misc_feature'].fillna('NA', inplace=True)

### Review null values in ordinal variables

In [38]:
# Show number of null values in ordinal variables with null values
var_null_count(df_train, ord_var)

bsmt_qual           55
bsmt_cond           55
bsmt_exposure       58
bsmtfin_type_1      55
bsmtfin_type_2      56
fireplace_qu       998
pool_qc           2039
fence             1648
dtype: int64

In [39]:
# Show number of null values in ordinal variables with null values
var_null_count(df_test, ord_var)

bsmt_qual          25
bsmt_cond          25
bsmt_exposure      25
bsmtfin_type_1     25
bsmtfin_type_2     25
electrical          1
fireplace_qu      422
pool_qc           875
fence             707
dtype: int64

In [40]:
# Display all unique categories in ordinal variables with null values 
unique_val_dict(df_train, ord_var)

{'bsmt_qual': array(['TA', 'Gd', 'Fa', nan, 'Ex', 'Po'], dtype=object),
 'bsmt_cond': array(['TA', 'Gd', nan, 'Fa', 'Po', 'Ex'], dtype=object),
 'bsmt_exposure': array(['No', 'Gd', 'Av', nan, 'Mn'], dtype=object),
 'bsmtfin_type_1': array(['GLQ', 'Unf', 'ALQ', 'Rec', nan, 'BLQ', 'LwQ'], dtype=object),
 'bsmtfin_type_2': array(['Unf', 'Rec', nan, 'BLQ', 'GLQ', 'LwQ', 'ALQ'], dtype=object),
 'fireplace_qu': array([nan, 'TA', 'Gd', 'Po', 'Ex', 'Fa'], dtype=object),
 'pool_qc': array([nan, 'Fa', 'Gd', 'Ex', 'TA'], dtype=object),
 'fence': array([nan, 'MnPrv', 'GdPrv', 'GdWo', 'MnWw'], dtype=object)}

In [41]:
# Display all unique categories in ordinal variables with null values 
unique_val_dict(df_test, ord_var)

{'bsmt_qual': array(['Fa', 'Gd', 'TA', 'Ex', nan, 'Po'], dtype=object),
 'bsmt_cond': array(['TA', 'Gd', nan, 'Fa'], dtype=object),
 'bsmt_exposure': array(['No', 'Av', nan, 'Mn', 'Gd'], dtype=object),
 'bsmtfin_type_1': array(['Unf', 'GLQ', 'BLQ', 'Rec', 'ALQ', nan, 'LwQ'], dtype=object),
 'bsmtfin_type_2': array(['Unf', 'LwQ', nan, 'ALQ', 'GLQ', 'Rec', 'BLQ'], dtype=object),
 'electrical': array(['FuseP', 'SBrkr', 'FuseA', 'FuseF', nan], dtype=object),
 'fireplace_qu': array([nan, 'Gd', 'Fa', 'TA', 'Po', 'Ex'], dtype=object),
 'pool_qc': array([nan, 'Ex', 'TA'], dtype=object),
 'fence': array([nan, 'MnPrv', 'GdPrv', 'GdWo', 'MnWw'], dtype=object)}

**Key Observations:**

From the data dictionary, we understand the following:
1. The NaNs in `bsmt_qual`, `bsmt_cond`, `bsmt_exposure`, `bsmtfin_type_1` and `bsmtfin_type_2` correspond to the property having no basement.
2. The NaNs in `fireplace_qu` correspond to no fireplace on the property.
3. The NaNs in `pool_qc` correspond to no pool on the property.
4. The NaNs in `fence` correspond to absence of fence on the property.
5. There is 1 NaN on `electrical` for the test set

#### Fix the null values in ordinal variables

Count number of properties where `bsmt_qual`, `bsmt_cond`, `bsmt_exposure`, `bsmtfin_type_1` and `bsmtfin_type_2` are all null

In [42]:
basement_count = df_train.loc[((((df_train['bsmt_qual'].isna())&
                            (df_train['bsmt_cond'].isna()))&
                            (df_train['bsmt_exposure'].isna()))&
                            (df_train['bsmtfin_type_1'].isna()))&
                            (df_train['bsmtfin_type_2'].isna())].shape[0] 
                            

print(f'There are {basement_count} properties where "bsmt_qual", "bsmt_cond", "bsmt_exposure", "bsmtfin_type_1" and "bsmtfin_type_2" are all null')



There are 55 properties where "bsmt_qual", "bsmt_cond", "bsmt_exposure", "bsmtfin_type_1" and "bsmtfin_type_2" are all null


In [43]:
basement_count = df_test.loc[((((df_test['bsmt_qual'].isna())&
                            (df_test['bsmt_cond'].isna()))&
                            (df_test['bsmt_exposure'].isna()))&
                            (df_test['bsmtfin_type_1'].isna()))&
                            (df_test['bsmtfin_type_2'].isna())].shape[0] 
                            

print(f'There are {basement_count} properties where "bsmt_qual", "bsmt_cond", "bsmt_exposure", "bsmtfin_type_1" and "bsmtfin_type_2" are all null')




There are 25 properties where "bsmt_qual", "bsmt_cond", "bsmt_exposure", "bsmtfin_type_1" and "bsmtfin_type_2" are all null


Show the properties where `bsmt_exposure` is null but `bsmt_qual` is not null

In [44]:
basement_list = ['bsmt_qual', 'bsmt_cond', 'bsmt_exposure', 'bsmtfin_type_1', 'bsmtfin_sf_1', 'bsmtfin_type_2', 
                 'bsmtfin_sf_2', 'bsmt_unf_sf', 'total_bsmt_sf']

df_train.loc[(df_train['bsmt_exposure'].isna())&
             (df_train['bsmt_qual'].notna()), basement_list]

bsmt_qual bsmt_cond bsmt_exposure bsmtfin_type_1  bsmtfin_sf_1  \
1456        Gd        TA           NaN            Unf           0.0   
1547        Gd        TA           NaN            Unf           0.0   
1997        Gd        TA           NaN            Unf           0.0   

     bsmtfin_type_2  bsmtfin_sf_2  bsmt_unf_sf  total_bsmt_sf  
1456            Unf           0.0        725.0          725.0  
1547            Unf           0.0       1595.0         1595.0  
1997            Unf           0.0        936.0          936.0

In [45]:
df_test.loc[(df_test['bsmt_exposure'].isna())&
             (df_test['bsmt_qual'].notna()), basement_list]

Empty DataFrame
Columns: [bsmt_qual, bsmt_cond, bsmt_exposure, bsmtfin_type_1, bsmtfin_sf_1, bsmtfin_type_2, bsmtfin_sf_2, bsmt_unf_sf, total_bsmt_sf]
Index: []

The above properties all has basements but somehow the `bsmt_exposure` were tagged as NaN. It is observed that these basement are in an unfinished state as seen from their corresponding tags in `bsmtfin_type_1` and `bsmtfin_type_2`. We will regard them has having no basement and tag them as 'NA'

In [46]:
df_train.loc[[1456,1547,1997], 'bsmt_exposure']='NA'

Show the properties where `bsmtfin_type_2` is null but `bsmt_qual` is not null

In [47]:
df_train.loc[(df_train['bsmtfin_type_2'].isna())&
             (df_train['bsmt_qual'].notna()), basement_list]

bsmt_qual bsmt_cond bsmt_exposure bsmtfin_type_1  bsmtfin_sf_1  \
1147        Gd        TA            No            GLQ        1124.0   

     bsmtfin_type_2  bsmtfin_sf_2  bsmt_unf_sf  total_bsmt_sf  
1147            NaN         479.0       1603.0         3206.0

In [48]:
df_test.loc[(df_test['bsmtfin_type_2'].isna())&
             (df_test['bsmt_qual'].notna()), basement_list]

Empty DataFrame
Columns: [bsmt_qual, bsmt_cond, bsmt_exposure, bsmtfin_type_1, bsmtfin_sf_1, bsmtfin_type_2, bsmtfin_sf_2, bsmt_unf_sf, total_bsmt_sf]
Index: []

This property will be removed as it has missing value on its bsmtfin_type_2 completely at random

In [49]:
df_train.drop(axis=0, index=[1147], inplace=True)

Replace all null values in `bsmt_qual`, `bsmt_cond`, `bsmt_exposure`, `bsmtfin_type_1`, and `bsmtfin_type_2`

In [50]:
for col_name in ['bsmt_qual', 'bsmt_cond', 'bsmt_exposure', 'bsmtfin_type_1', 'bsmtfin_type_2']:
    df_train[col_name].fillna('NA', inplace=True)

In [51]:
for col_name in ['bsmt_qual', 'bsmt_cond', 'bsmt_exposure', 'bsmtfin_type_1', 'bsmtfin_type_2']:
    df_test[col_name].fillna('NA', inplace=True)

In [52]:
# Show number of null values in ordinal variables with null values
var_null_count(df_train, ord_var)

fireplace_qu     998
pool_qc         2038
fence           1647
dtype: int64

In [53]:
# Show number of null values in ordinal variables with null values
var_null_count(df_test, ord_var)

electrical        1
fireplace_qu    422
pool_qc         875
fence           707
dtype: int64

`fence` is a standalone variable which the null values can be replaced easily without considering other variables

In [54]:
df_train['fence'].fillna('NA', inplace=True)

In [55]:
df_test['fence'].fillna('NA', inplace=True)

We can replace the null values in `fireplace_qu` and `pool_qc` with 'NA' as it means there are no fireplace and pool respectively

In [56]:
df_train['fireplace_qu'].fillna('NA', inplace=True)
df_train['pool_qc'].fillna('NA', inplace=True)

In [57]:
df_test['fireplace_qu'].fillna('NA', inplace=True)
df_test['pool_qc'].fillna('NA', inplace=True)

In [58]:
df_test.loc[df_test['electrical'].isna(), :]

id        pid  ms_subclass ms_zoning  lot_frontage  lot_area street  \
635  1578  916386080           80        RL          73.0      9735   Pave   

    alley lot_shape land_contour utilities lot_config land_slope neighborhood  \
635    NA       Reg          Lvl    AllPub     Inside        Gtl       Timber   

    condition_1 condition_2 bldg_type house_style  overall_qual  overall_cond  \
635        Norm        Norm      1Fam        SLvl             5             5   

     year_built  year_remod_add roof_style roof_matl exterior_1st  \
635        2006            2007      Gable   CompShg      VinylSd   

    exterior_2nd mas_vnr_type  mas_vnr_area exter_qual exter_cond foundation  \
635      VinylSd         None           0.0         TA         TA      PConc   

    bsmt_qual bsmt_cond bsmt_exposure bsmtfin_type_1  bsmtfin_sf_1  \
635        Gd        TA            No            Unf             0   

    bsmtfin_type_2  bsmtfin_sf_2  bsmt_unf_sf  total_bsmt_sf heating  \
635            Unf             0          384            384    GasA   

    heating_qc central_air electrical  1st_flr_sf  2nd_flr_sf  \
635         Gd           Y        NaN         754         640   

     low_qual_fin_sf  gr_liv_area  bsmt_full_bath  bsmt_half_bath  full_bath  \
635                0         1394               0               0          2   

     half_bath  bedroom_abvgr  kitchen_abvgr kitchen_qual  totrms_abvgrd  \
635          1              3              1           Gd              7   

    functional  fireplaces fireplace_qu garage_type garage_yr_blt  \
635        Typ           0           NA     BuiltIn        2007.0   

    garage_finish  garage_cars  garage_area garage_qual garage_cond  \
635           Fin            2          400          TA          TA   

    paved_drive  wood_deck_sf  open_porch_sf  enclosed_porch  3ssn_porch  \
635           Y           100              0               0           0   

     screen_porch  pool_area pool_qc fence misc_feature  misc_val  mo_sold  \
635             0          0      NA    NA           NA         0        5   

     yr_sold sale_type  
635     2008       WD

Electrical on the test set seems to be missing completely at random. We will impute with the most common category

In [59]:
df_test['electrical'].fillna(df_test['electrical'].value_counts().index[0], inplace=True)


In [60]:
# Show number of null values in ordinal variables with null values
var_null_count(df_train, ord_var)

Series([], dtype: int64)

In [61]:
# Show number of null values in ordinal variables with null values
var_null_count(df_test, ord_var)

Series([], dtype: int64)

All ordinal variables do not have null values

### Review null values in discrete and continuous variables

In [62]:
# Display number of null values in discrete variables
var_null_count(df_train, disc_var)


bsmt_full_bath    2
bsmt_half_bath    2
garage_cars       1
dtype: int64

In [63]:
# Display number of null values in discrete variables
var_null_count(df_test, disc_var)

Series([], dtype: int64)

In [64]:
# Display number of null values in continuous variables
var_null_count(df_train, cont_var)

lot_frontage     330
bsmtfin_sf_1       1
bsmtfin_sf_2       1
bsmt_unf_sf        1
total_bsmt_sf      1
garage_area        1
dtype: int64

In [65]:
# Display number of null values in continuous variables
var_null_count(df_test, cont_var)

lot_frontage    160
dtype: int64

In [66]:
# Display all unique categories in discrete variables with null values 
unique_val_dict(df_train, disc_var)

{'bsmt_full_bath': array([ 0.,  1.,  2., nan,  3.]),
 'bsmt_half_bath': array([ 0.,  1., nan,  2.]),
 'garage_cars': array([ 2.,  1.,  3.,  0.,  4.,  5., nan])}

**Key Observations:**

1. There are NaNs in both `bsmt_full_bath` and `bsmt_half_bath`. We need to look at the rows in which they reside to make a decision on what to do with them.
2. There are NaNs in `bsmtfin_sf_1`, `bsmtfin_sf_2`, `bsmt_unf_sf` and `total_bsmt_sf`. We have to check if they correspond to properties without basements
3. There are 330 NaNs for `lot_frontage`. The missing values will need to be imputed as the number of missing values is too large.
4. For the test set, there are 160 NaNs for lot_frontage for the continuous variables and none for the discrete variables

#### Fix the null values in continuous and discrete variables

Show the properties which has NaNs for `bsmt_full_bath` and `bsmt_half_bath`.

In [67]:
df_train.loc[df_train['bsmt_full_bath'].isna(), :]

id        pid  ms_subclass ms_zoning  lot_frontage  lot_area street  \
616   1498  908154080           20        RL         123.0     47007   Pave   
1327  1342  903230120           20        RM          99.0      5940   Pave   

     alley lot_shape land_contour utilities lot_config land_slope  \
616     NA       IR1          Lvl    AllPub     Inside        Gtl   
1327    NA       IR1          Lvl    AllPub        FR3        Gtl   

     neighborhood condition_1 condition_2 bldg_type house_style  overall_qual  \
616       Edwards        Norm        Norm      1Fam      1Story             5   
1327      BrkSide       Feedr        Norm      1Fam      1Story             4   

      overall_cond  year_built  year_remod_add roof_style roof_matl  \
616              7        1959            1996      Gable   CompShg   
1327             7        1946            1950      Gable   CompShg   

     exterior_1st exterior_2nd mas_vnr_type  mas_vnr_area exter_qual  \
616       Plywood      Plywood         None           0.0         TA   
1327      MetalSd       CBlock         None           0.0         TA   

     exter_cond foundation bsmt_qual bsmt_cond bsmt_exposure bsmtfin_type_1  \
616          TA       Slab        NA        NA            NA             NA   
1327         TA      PConc        NA        NA            NA             NA   

      bsmtfin_sf_1 bsmtfin_type_2  bsmtfin_sf_2  bsmt_unf_sf  total_bsmt_sf  \
616            0.0             NA           0.0          0.0            0.0   
1327           NaN             NA           NaN          NaN            NaN   

     heating heating_qc central_air electrical  1st_flr_sf  2nd_flr_sf  \
616     GasA         TA           Y      SBrkr        3820           0   
1327    GasA         TA           Y      FuseA         896           0   

      low_qual_fin_sf  gr_liv_area  bsmt_full_bath  bsmt_half_bath  full_bath  \
616                 0         3820             NaN             NaN          3   
1327                0          896             NaN             NaN          1   

      half_bath  bedroom_abvgr  kitchen_abvgr kitchen_qual  totrms_abvgrd  \
616           1              5              1           Ex             11   
1327          0              2              1           TA              4   

     functional  fireplaces fireplace_qu garage_type garage_yr_blt  \
616         Typ           2           Gd      Attchd        1959.0   
1327        Typ           0           NA      Detchd        1946.0   

     garage_finish  garage_cars  garage_area garage_qual garage_cond  \
616            Unf          2.0        624.0          TA          TA   
1327           Unf          1.0        280.0          TA          TA   

     paved_drive  wood_deck_sf  open_porch_sf  enclosed_porch  3ssn_porch  \
616            Y             0            372               0           0   
1327           Y             0              0               0           0   

      screen_porch  pool_area pool_qc  fence misc_feature  misc_val  mo_sold  \
616              0          0      NA     NA           NA         0        7   
1327             0          0      NA  MnPrv           NA         0        4   

      yr_sold sale_type  saleprice  
616      2008       WD      284700  
1327     2008     ConLD      79000

There are no basement for the above properties. The NaNs in `bsmtfin_sf_1`, `bsmtfin_sf_2`, `bsmt_unf_sf`, `total_bsmt_sf` on property id 1342 will be imputed with 0 to align with id 1498. `bsmt_full_bath` and `bsmt_half_bath` will also be imputed with 0. 

In [68]:
for col_name in ['bsmt_full_bath', 'bsmt_half_bath', 'bsmtfin_sf_1', 'bsmtfin_sf_2', 'bsmt_unf_sf', 'total_bsmt_sf']:
    df_train[col_name].fillna(0, inplace=True)

In [69]:
# Display number of null values in discrete variables
var_null_count(df_train, disc_var)

garage_cars    1
dtype: int64

All discrete variables do not contain null values

In [70]:
# Display number of null values in continuous variables
var_null_count(df_train, cont_var)

lot_frontage    330
garage_area       1
dtype: int64

In [71]:
df_train.loc[df_train['garage_cars'].isnull()]

id        pid  ms_subclass ms_zoning  lot_frontage  lot_area street  \
1712  2237  910201180           70        RM          50.0      9060   Pave   

     alley lot_shape land_contour utilities lot_config land_slope  \
1712    NA       Reg          Lvl    AllPub     Inside        Gtl   

     neighborhood condition_1 condition_2 bldg_type house_style  overall_qual  \
1712       IDOTRR        Norm        Norm      1Fam      2Story             5   

      overall_cond  year_built  year_remod_add roof_style roof_matl  \
1712             6        1923            1999      Gable   CompShg   

     exterior_1st exterior_2nd mas_vnr_type  mas_vnr_area exter_qual  \
1712      Wd Sdng      Plywood         None           0.0         TA   

     exter_cond foundation bsmt_qual bsmt_cond bsmt_exposure bsmtfin_type_1  \
1712         TA     BrkTil        Gd        TA            No            ALQ   

      bsmtfin_sf_1 bsmtfin_type_2  bsmtfin_sf_2  bsmt_unf_sf  total_bsmt_sf  \
1712         548.0            Unf           0.0        311.0          859.0   

     heating heating_qc central_air electrical  1st_flr_sf  2nd_flr_sf  \
1712    GasA         Ex           Y      SBrkr         942         886   

      low_qual_fin_sf  gr_liv_area  bsmt_full_bath  bsmt_half_bath  full_bath  \
1712                0         1828             0.0             0.0          2   

      half_bath  bedroom_abvgr  kitchen_abvgr kitchen_qual  totrms_abvgrd  \
1712          0              3              1           Gd              6   

     functional  fireplaces fireplace_qu garage_type garage_yr_blt  \
1712        Typ           0           NA      Detchd            NA   

     garage_finish  garage_cars  garage_area garage_qual garage_cond  \
1712            NA          NaN          NaN          NA          NA   

     paved_drive  wood_deck_sf  open_porch_sf  enclosed_porch  3ssn_porch  \
1712           Y           174              0             212           0   

      screen_porch  pool_area pool_qc  fence misc_feature  misc_val  mo_sold  \
1712             0          0      NA  MnPrv           NA         0        3   

      yr_sold sale_type  saleprice  
1712     2007       WD      150909

The property with both `garage_cars` and `garage_area` having null values are the same. We will imputboth `garage_area` and `garage_cars` as 0

In [72]:
df_train['garage_cars'].fillna(0, inplace=True)
df_train['garage_area'].fillna(0, inplace=True)

### Data Cleaned

All null variables has been removed on both the train and test sets except for `lot_frontage`. This will be done in the next notebook for EDA

In [73]:
df_train.isna().sum()

id                   0
pid                  0
ms_subclass          0
ms_zoning            0
lot_frontage       330
lot_area             0
street               0
alley                0
lot_shape            0
land_contour         0
utilities            0
lot_config           0
land_slope           0
neighborhood         0
condition_1          0
condition_2          0
bldg_type            0
house_style          0
overall_qual         0
overall_cond         0
year_built           0
year_remod_add       0
roof_style           0
roof_matl            0
exterior_1st         0
exterior_2nd         0
mas_vnr_type         0
mas_vnr_area         0
exter_qual           0
exter_cond           0
foundation           0
bsmt_qual            0
bsmt_cond            0
bsmt_exposure        0
bsmtfin_type_1       0
bsmtfin_sf_1         0
bsmtfin_type_2       0
bsmtfin_sf_2         0
bsmt_unf_sf          0
total_bsmt_sf        0
heating              0
heating_qc           0
central_air          0
electrical 

In [74]:
df_test.isna().sum()

id                   0
pid                  0
ms_subclass          0
ms_zoning            0
lot_frontage       160
lot_area             0
street               0
alley                0
lot_shape            0
land_contour         0
utilities            0
lot_config           0
land_slope           0
neighborhood         0
condition_1          0
condition_2          0
bldg_type            0
house_style          0
overall_qual         0
overall_cond         0
year_built           0
year_remod_add       0
roof_style           0
roof_matl            0
exterior_1st         0
exterior_2nd         0
mas_vnr_type         0
mas_vnr_area         0
exter_qual           0
exter_cond           0
foundation           0
bsmt_qual            0
bsmt_cond            0
bsmt_exposure        0
bsmtfin_type_1       0
bsmtfin_sf_1         0
bsmtfin_type_2       0
bsmtfin_sf_2         0
bsmt_unf_sf          0
total_bsmt_sf        0
heating              0
heating_qc           0
central_air          0
electrical 

Both the test and train data is saved and processed in the next Notebook: EDA and Features Engineering

In [75]:
df_train.to_csv('../data/train_cleaned.csv', index=False)
df_test.to_csv('../data/test_cleaned.csv', index=False)